marimbabot_vision/scripts/train.py

```diff
@@ -0,0 +1,221 @@
+        dataset_path: str,
+        max_length: int,
+        split: str = "train",
+        ignore_id: int = -100,
+        start_token: str = "<s>"
+    ):
+        super().__init__()
+
+        self.max_length = max_length
+        self.split = split
+        self.ignore_id = ignore_id
+        self.start_token = start_token
+
+        self.dataset_lilypond_files = glob.glob(f"{dataset_path}/*/*.txt")
+        self.dataset_length = len(self.dataset_lilypond_files)
+
+        self.gt_token_sequences = []
+        for sample in self.dataset_lilypond_files:
+            with open(sample, 'r') as f:
+                ground_truth = f.read()
+            self.gt_token_sequences.append(
+                    ground_truth + pre_processor.tokenizer.eos_token
+            )
+
+        self.add_tokens([self.start_token])
+
+    def add_tokens(self, list_of_tokens: List[str]):
+        newly_added_num = pre_processor.tokenizer.add_tokens(list_of_tokens)
+        if newly_added_num > 0:
+            model.decoder.resize_token_embeddings(len(pre_processor.tokenizer))
+    
+    def __len__(self) -> int:
+        return self.dataset_length
+
+    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        sample = self.dataset_lilypond_files[idx]
+        
+        image = Image.open(glob.glob(f"{os.path.dirname(sample)}/*.png")[0])
+
+        pixel_values = pre_processor(image, random_padding=self.split == "train", return_tensors="pt").pixel_values
+        pixel_values = pixel_values.squeeze()
+
+        # Targets
+        target_sequence = self.gt_token_sequences[idx]
+        input_ids = pre_processor.tokenizer(
+            target_sequence,
+            add_special_tokens=False,
+            max_length=self.max_length,
+            padding="max_length",
+            truncation=True,
+            return_tensors="pt",
+        )["input_ids"].squeeze(0)
+
+        labels = input_ids.clone()
+        labels[labels == pre_processor.tokenizer.pad_token_id] = self.ignore_id
+        return pixel_values, labels, target_sequence
+
+# Instantiate dataset and dataloader
+
+train_dataset = NoteDataset(config['train_data_path'], max_length=config['max_length'], split="train", start_token=config['start_token'])
+val_dataset = NoteDataset(config['val_data_path'], max_length=config['max_length'], split="validation", start_token=config['start_token'])
+
+train_dataloader = DataLoader(train_dataset, batch_size=config['train_batch_sizes'][0], shuffle=True, num_workers=config['num_workers'])
+val_dataloader = DataLoader(val_dataset, batch_size=config['val_batch_sizes'][0], shuffle=False, num_workers=config['num_workers'])
+
+
+# Create pytorch lightning module
+class DonutModelPLModule(pl.LightningModule):
+    def __init__(self, config, pre_processor, model):
+        super().__init__()
+        self.config = config
+        self.pre_processor = pre_processor
+        self.model = model
+
+    def training_step(self, batch, batch_idx):
```

Comment:
a "step" equals one batch here?

Reply:
Yes, we process a single batch each step. 

Also see: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html